rfd/0054-passwordless-macos.md

```diff
@@ -0,0 +1,419 @@
+  the more restrictive of the possible "Accessibility Values".
+
+* [kSecAccessControlBiometryAny](
+  https://developer.apple.com/documentation/security/secaccesscontrolcreateflags/ksecaccesscontrolbiometryany)
+  requires a biometric check (Touch ID on macOS). It is more restrictive than
+  *kSecAccessControlUserPresence* (which allows passwords), but less restrictive
+  than *kSecAccessControlBiometryCurrentSet* (doesn't work with newly enrolled
+  fingerprints). *kSecAccessControlBiometryAny* seems to be the sweet spot of
+  security and usability.
+
+In case of a registration failure, `tsh` must do its best to delete the
+created-but-not-registered credential. If all fails, it is possible to use the
+hidden [`tsh` support commands](#tsh-support-commands) for a manual cleanup.
+
+### Authentication
+
+Authentication offers a plethora of options, depending on both server settings
+(otp, webauthn, passwordless) and client state (FIDO2 keys present, Touch ID
+keys registered). In order to decide which flow to follow, `tsh` must first
+assess what is possible, preferably without asking for unnecessary user
+interaction.
+
+Unlike FIDO2 keys, it is possible for `tsh` to discover if Touch ID keys are
+registered in the Enclave _without_ user interaction. Because all Touch ID keys
+are functionally resident keys, as long as the server supports passwordless,
+then `tsh` is free to use it.
+
+If Touch ID passwordless authentication is possible, then it is the preferred
+authentication mode and is used by default.
+
+If passwordless is not possible, but Touch ID is present, then Touch ID MFA is
+also preferred and used by default.
+
+To allow users agency over the eager behaviors of Touch ID, `tsh` is augmented
+with the following flags:
+
+* `tsh --pwdless_mode={auto,on,off}` - activate/deactivate passwordless
+  authentication
+
+    `auto` is the default behavior described above (passwordless is preferred
+    if we are certain it may be used)
+
+    `on` enables passwordless logins (`tsh login --pwdless`, described in the
+    [Passwordless FIDO2 RFD][passwordless fido2], is a shortcut to it)
+
+    `off` disables passwordless logins
+
+* `tsh --mfa_mode={auto,platform,cross-platform}` - choose whether to use
+  platform or cross-platform MFA
+
+    `auto` is the default behavior described above, which favors Touch ID
+
+    `platform` prefers platform authenticators, such as Touch ID, over OTP or
+    portable FIDO2 keys
+
+    `cross-platform` prefers FIDO2 or OTP (aka `tsh` behavior prior to this RFD)
+
+Both `--pwdless_mode` and `--mfa_mode` may be specified for fine-grained control
+over `tsh` authentication logic. Note that those flags apply not only to
+`tsh login`, but also for any commands that require re-authentication (such as
+`tsh mfa add`).
+
+Finally, if there are Touch ID credentials for multiple users and the login user
+is not known, `tsh login` may prompt the user to specify the `--user` flag.
```

Comment:
That was actually my first draft, more below.

> From a security perspective, we show the TouchID user all available users tied to their TouchID credential. That seems fine though because all these users do belong to the TouchID user.

The information leakage is what prompted me to change it to the more "obfuscated" version. I'm not concerned about the legitimate user, but if a third party that gets their hands on a (unlocked) macOS they could see which credentials it holds by doing `tsh login --proxy=example.com` and cancelling after the prompt. It's not huge, but I'd rather sidestep it if I can.

Reply:
Understood. The risk, from a UX perspective, is that the user might not recall the exact login they want to use. What are some of the easiest ways that a user can recall their login? Perhaps reverse-i-search and checking the web app?